### PR TITLE
Add fittable

### PR DESCRIPTION
**Pull request by stephantul**

### CODE REVIEW

File: model2vec/train/classifier.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    def predict_proba(self, X: list[str]) -> np.ndarray:
+        """Predict the probability of each class."""
+        pred: list[np.ndarray] = []
+        for batch in range(0, len(X), 1024):
+            logits = self._predict(X[batch : batch + 1024])
+            pred.append(torch.softmax(logits, dim=1).numpy())
+
+        return np.concatenate(pred)
+
+    def fit(
```
Comment (Pringled) [inline model2vec/train/classifier.py:87]: This function is quite long and complex, I think it would be better/more readable to split this into smaller parts, but up to you

File: model2vec/train/base.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        embedded = torch.bmm(w[:, None, :], embedded).squeeze(1)
+        # embedded = embedded.sum(1)
+        embedded = embedded / length[:, None]
+
+        return nn.functional.normalize(embedded)
+
+    def forward(self, input_ids: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
+        """Forward pass through the mean, and a classifier layer after."""
+        encoded = self._encode(input_ids)
+        return self.head(encoded), encoded
```
Comment (Pringled) [inline model2vec/train/base.py:90]: Is the second tensor (encoded) actually used anywhere in the code?

File: pyproject.toml [hunk 3]
```diff
@@ -54,8 +54,10 @@ dev = [
     "pytest-cov",
     "ruff",
 ]
+
 distill = ["torch", "transformers", "scikit-learn"]
 onnx = ["onnx", "torch"]
+train = ["torch", "lightning"]
```
Comment (Pringled) [inline pyproject.toml:60]: I think it's a bit safer to pin these (at least softpin) as torch and lightning tend to have breaking major releases, but we are also not doing that in distill, wdyt? Maybe just < 3 would make sense

File: model2vec/train/classifier.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    def __init__(
+        self,
+        *,
+        vectors: torch.Tensor,
+        tokenizer: Tokenizer,
+        n_layers: int,
+        hidden_dim: int,
+        out_dim: int,
+        pad_id: int = 0,
+    ) -> None:
```
Comment (Pringled) [inline model2vec/train/classifier.py:31]: Imo learning rate should also be a changeable parameter

File: model2vec/train/classifier.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+        self.head = self.construct_head()
+        self.embeddings = nn.Embedding.from_pretrained(self.vectors.clone(), freeze=False, padding_idx=self.pad_id)
+
+        label_mapping = {label: idx for idx, label in enumerate(self.classes)}
+        label_counts = Counter(y)
+        if min(label_counts.values()) < 2:
+            logger.info("Some classes have less than 2 samples. Stratification is disabled.")
+            train_texts, validation_texts, train_labels, validation_labels = train_test_split(
+                X, y, test_size=0.1, random_state=42, shuffle=True
```
Comment (Pringled) [inline model2vec/train/classifier.py:109]: Should we allow for more control over train/test splitting (e.g. train/test/val splits, setting the sizes, etc), or is that something users can edit themselves here if desired?

File: model2vec/train/classifier.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+        val_tokenized: list[list[int]] = [
+            encoding.ids for encoding in self.tokenizer.encode_batch_fast(validation_texts, add_special_tokens=False)
+        ]
+        val_labels_tensor = torch.Tensor([label_mapping[label] for label in validation_labels]).long()
+        val_dataset = TextDataset(val_tokenized, val_labels_tensor)
+
+        c = ClassifierLightningModule(self)
+
+        batch_size = 32
```
Comment (Pringled) [inline model2vec/train/classifier.py:131]: Would make this a default param

**Review by Pringled — CHANGES_REQUESTED** [review]
Looks good! Some minor comments and suggestions.

**Review by Pringled — APPROVED** [review]

File: model2vec/inference/model.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        model, head = _load_pipeline(path, token)
+        model.embedding = np.nan_to_num(model.embedding)
+
+        return cls(model, head)
+
+    def save_pretrained(self, path: str) -> None:
+        """Save the model to a folder."""
+        save_pipeline(self, path)
+
+    def push_to_hub(self, repo_id: str, token: str | None = None, private: bool = False) -> None:
```
Comment (davidberenstein1957) [inline model2vec/inference/model.py:45]: I would add a modelcard and perhaps tags or a library reference, this helps a lot with visibility, usability and findability.

https://huggingface.co/docs/hub/model-cards#specifying-a-library

File: model2vec/inference/model.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+
+class StaticModelPipeline:
+    def __init__(self, model: StaticModel, head: Pipeline) -> None:
+        """Create a pipeline with a StaticModel encoder."""
+        self.model = model
+        self.head = head
+
+    @classmethod
+    def from_pretrained(
```
Comment (davidberenstein1957) [inline model2vec/inference/model.py:26]: can't we load it from the Hub? perhaps we should align the arguments a bit with the `transformers` naming given you've also adopted `from_pretrained`? 

For example using `pretrained_model_name_or_path`. https://huggingface.co/docs/transformers/v4.48.0/en/model_doc/auto#transformers.AutoTokenizer.from_pretrained

File: model2vec/inference/model.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        """Predict the labels of the input."""
+        encoded = self._predict_and_coerce_to_2d(X)
+
+        return self.head.predict(encoded)
+
+    def predict_proba(self, X: list[str] | str) -> np.ndarray:
+        """Predict the probabilities of the labels of the input."""
+        encoded = self._predict_and_coerce_to_2d(X)
+
+        return self.head.predict_proba(encoded)
```
Comment (davidberenstein1957) [inline model2vec/inference/model.py:78]: Maybe we can also add a `__call__`?

File: model2vec/model.py [hunk 10]
```diff
@@ -66,7 +66,8 @@ def __init__(
             self._can_encode_fast = False
 
         if normalize is not None:
-            self.normalize = normalize
+            self._normalize = normalize
+            self.config["normalize"] = normalize
         else:
             self.normalize = self.config.get("normalize", False)
```
Comment (davidberenstein1957) [inline model2vec/model.py:73]: Perhaps we can add `get_sentence_embedding_dimension()` like sentence transformers?

File: model2vec/train/README.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
+# Training
+
+Aside from [distillation](../../README.md#distillation), `model2vec` also supports training simple classifiers on top of static models, using [pytorch](https://pytorch.org/) and [lightning](https://lightning.ai/).
```
Comment (davidberenstein1957) [inline model2vec/train/README.md:3]: And sklearn right?

File: model2vec/train/classifier.py [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+        X: list[str], y: list[str], test_size: float
+    ) -> tuple[list[str], list[str], list[str], list[str]]:
+        """Split the data."""
+        label_counts = Counter(y)
+        if min(label_counts.values()) < 2:
+            logger.info("Some classes have less than 2 samples. Stratification is disabled.")
+            return train_test_split(X, y, test_size=test_size, random_state=42, shuffle=True)
+        return train_test_split(X, y, test_size=test_size, random_state=42, shuffle=True, stratify=y)
+
+    def to_pipeline(self) -> StaticModelPipeline:
```
Comment (davidberenstein1957) [inline model2vec/train/classifier.py:219]: Should we be mre epxlicit and convert call this `to_sklearn_pipeline`?

File: model2vec/train/classifier.py [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+        return vectors
+
+    def predict_proba(self, X: list[str], show_progress_bar: bool = False, batch_size: int = 1024) -> np.ndarray:
+        """Predict the probability of each class."""
+        pred: list[np.ndarray] = []
+        for batch in trange(0, len(X), batch_size, disable=not show_progress_bar):
+            logits = self._predict_single_batch(X[batch : batch + batch_size])
+            pred.append(torch.softmax(logits, dim=1).numpy())
+
+        return np.concatenate(pred)
```
Comment (davidberenstein1957) [inline model2vec/train/classifier.py:89]: Do we want a `__call__` too?

File: model2vec/train/README.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+# Persistence
+
+You can turn a classifier into a scikit-learn compatible pipeline, as follows:
+
+```python
+pipeline = classifier.to_pipeline()
+```
+
+This pipeline object can be persisted using standard pickle-based methods, such as [joblib](https://joblib.readthedocs.io/en/stable/). This makes it easy to use your model in inferene pipelines (no installing torch!).
```
Comment (davidberenstein1957) [inline model2vec/train/README.md:76]: I believe there are some security concerns with pickle and joblib https://calmcode.io/course/scikit-save/security.

File: model2vec/train/README.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+Loading pipelines in this way is _extremely_ fast. It takes only 30ms to load a pipeline from disk.
+
+# Results
+
+The main results are detailed in our training blogpost, but we'll do a comparison with vanilla model2vec here. In a vanilla model2vec classifier, you just put a scikit-learn `LogisticRegressionCV` on top of the model encoder. In contrast, training a `StaticModelForClassification` fine-tunes the full model, including the `StaticModel` weights.
+
+We use 14 classification datasets, using 1000 examples from the train set, and the full test set. No parameters were tuned on any validation set. All datasets were taken from the [Setfit organization on Hugging Face](https://huggingface.co/datasets/SetFit).
+
+| dataset name               |   logistic regression head   |  full finetune |
```
Comment (davidberenstein1957) [inline model2vec/train/README.md:101]: Perhaps show how it compares to setfit, it being a direct competitor?

File: model2vec/train/classifier.py [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+    def construct_head(self) -> nn.Sequential:
+        """Constructs a simple classifier head."""
+        if self.n_layers == 0:
+            return nn.Sequential(nn.Linear(self.embed_dim, self.out_dim))
+        modules = [
+            nn.Linear(self.embed_dim, self.hidden_dim),
+            nn.ReLU(),
+        ]
+        for _ in range(self.n_layers - 1):
+            modules.extend([nn.Linear(self.hidden_dim, self.hidden_dim), nn.ReLU()])
```
Comment (davidberenstein1957) [inline model2vec/train/classifier.py:57]:
```suggestion
            modules.extend([nn.Linear(self.hidden_dim, self.hidden_dim), nn.ReLU()] * (self.n_layers - 1))
```

File: model2vec/train/classifier.py [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+        :param learning_rate: The learning rate.
+        :param batch_size: The batch size.
+            If this is None, a good batch size is chosen automatically.
+        :param early_stopping_patience: The patience for early stopping.
+            If this is None, early stopping is disabled.
+        :param test_size: The test size for the train-test split.
+        :param device: The device to train on. If this is "auto", the device is chosen automatically.
+        :return: The fitted model.
+        """
+        pl.seed_everything(42)
```
Comment (davidberenstein1957) [inline model2vec/train/classifier.py:121]: Don't we want to be able to pass a seed?

File: model2vec/train/classifier.py [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+        # If the dataset is small, we check the validation set every epoch.
+        # If the dataset is large, we check the validation set every 250 batches.
+        if n_train_batches < 250:
+            val_check_interval = None
+            check_val_every_epoch = 1
+        else:
+            val_check_interval = max(250, 2 * len(val_dataset) // batch_size)
+            check_val_every_epoch = None
+
+        with TemporaryDirectory() as tempdir:
```
Comment (davidberenstein1957) [inline model2vec/train/classifier.py:155]: This is to not create a random training directory when fitting?

File: model2vec/train/classifier.py [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+        if min(label_counts.values()) < 2:
+            logger.info("Some classes have less than 2 samples. Stratification is disabled.")
+            return train_test_split(X, y, test_size=test_size, random_state=42, shuffle=True)
+        return train_test_split(X, y, test_size=test_size, random_state=42, shuffle=True, stratify=y)
+
+    def to_pipeline(self) -> StaticModelPipeline:
+        """Convert the model to an sklearn pipeline."""
+        static_model = self.to_static_model()
+
+        random_state = np.random.RandomState(42)
```
Comment (davidberenstein1957) [inline model2vec/train/classifier.py:223]: same w.r.t. random seed. If we don't want users to be able to define this. Perhaps we could create `RANDOM_SEED` somewhere and re-use it instead of copy-pasting 42?

**Review by davidberenstein1957 — COMMENTED** [review]
Looks super useful. left some comments. Also, perhaps we can add some reference to multi-label usage somewhere?

File: model2vec/model.py [hunk 20]
```diff
@@ -66,7 +66,8 @@ def __init__(
...
```
[the remark this answers is not in the document]
Comment (stephantul) [reply model2vec/model.py:72]: This is intentional. `normalize` checks whether the value had been set before in the config and warns the user if it was something other than the provided value. assigning to `self._normalize` just sets it directly.

[the remark this answers is not in the document]
Comment (stephantul) [reply model2vec/model.py:72]: I do think we could change the setter to fix this.

File: model2vec/inference/model.py (hunk 7 above)
Comment (stephantul) [reply model2vec/inference/model.py:45]: This actually already happens because we push the underlying static model to the hub, which has a model card. This model card template is specified in the root of the code.

File: model2vec/inference/model.py (hunk 8 above)
Comment (stephantul) [reply model2vec/inference/model.py:26]: from_pretrained loads from the hub. The arguments mimic the ones from `StaticModel` and, although they don't match transformers exactly, we're wary of introducing breaking changes.

File: model2vec/model.py (hunk 10 above)
Comment (stephantul) [reply model2vec/model.py:73]: this is already in here, it's called `dim`. It's a property, not a function.

File: model2vec/train/README.md (hunk 11 above)
Comment (stephantul) [reply model2vec/train/README.md:3]: Good catch!

File: model2vec/train/classifier.py (hunk 12 above)
Comment (stephantul) [reply model2vec/train/classifier.py:219]: I thought of this, but it's actually not an sklearn pipeline (as in `sklearn.pipeline.Pipeline`). Maybe `to_staticmodelpipeline`?

File: model2vec/train/README.md (hunk 14 above)
Comment (stephantul) [reply model2vec/train/README.md:76]: We load using `skops` internally, so pushing to the hub and so on is safe if you use this library. But yeah I'll add a note to this effect.

File: model2vec/train/classifier.py (hunk 18 above)
Comment (stephantul) [reply model2vec/train/classifier.py:155]: yep! We do need a training directory to be able to load the best checkpoint.

File: model2vec/inference/README.md [hunk 21]
```diff
@@ -0,0 +1,18 @@
+# Inference
+
+This subpackage mainly contains helper functions for inference with trained models that have been exported to `scikit-learn` compatible pipelines.
+
+If you're looking for information on how to train a model, see [here](../train/README.md).
+
+# Usage
+
+Let's assume you're using our `potion-edu classifier`.
```
Comment (Pringled) [inline model2vec/inference/README.md:9]:
```suggestion
Let's assume you're using our [potion-edu classifier](https://huggingface.co/minishlab/potion-8m-edu-classifier).
```

File: model2vec/inference/README.md [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+
+```python
+from model2vec.inference import StaticModelPipeline
+
+s = StaticModelPipeline.from_pretrained("minishlab/potion-8m-edu-classifier")
```
Comment (Pringled) [inline model2vec/inference/README.md:14]:
```suggestion
classifier = StaticModelPipeline.from_pretrained("minishlab/potion-8m-edu-classifier")
```

File: model2vec/inference/README.md [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+label = s.predict("Attitudes towards cattle in the Alps: a study in letting go.")
```
Comment (Pringled) [inline model2vec/inference/README.md:15]:
```suggestion
label = classifier.predict("Attitudes towards cattle in the Alps: a study in letting go.")
```

File: model2vec/inference/model.py [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+
+class StaticModelPipeline:
+    def __init__(self, model: StaticModel, head: Pipeline) -> None:
+        """Create a pipeline with a StaticModel encoder."""
+        self.model = model
+        self.head = head
+
+    @classmethod
+    def from_pretrained(
+        cls: type[StaticModelPipeline], path: PathLike, token: str | None = None
```
Comment (Pringled) [inline model2vec/inference/model.py:27]: Should also accept trust_remote_code, and pass it to _load_pipeline I think

File: model2vec/inference/model.py [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+        from model2vec.hf_utils import push_folder_to_hub
+
+        with TemporaryDirectory() as temp_dir:
+            save_pipeline(self, temp_dir)
+            self.model.save_pretrained(temp_dir)
+            push_folder_to_hub(Path(temp_dir), repo_id, private, token)
+
+    def _predict_and_coerce_to_2d(self, X: list[str] | str) -> np.ndarray:
+        """Predict the labels of the input and coerce the output to a matrix."""
+        encoded = self.model.encode(X)
```
Comment (Pringled) [inline model2vec/inference/model.py:64]: There's no control over encode since this is in a private function, I'm not sure if we want to give any control here? E.g batch size, multiprocessing, etc?

File: model2vec/train/classifier.py [hunk 26]
```diff
@@ -0,0 +1,297 @@
...
+        accuracy = (head_out.argmax(1) == y).float().mean()
+
+        self.log("val_loss", loss)
+        self.log("val_accuracy", accuracy, prog_bar=True)
+
+        return loss
+
+    def configure_optimizers(self) -> OptimizerLRScheduler:
+        """Simple Adam optimizer."""
+        optimizer = torch.optim.Adam(self.model.parameters(), lr=1e-3)
```
Comment (Pringled) [inline model2vec/train/classifier.py:285]: I think learning rate is now ignored, this should be:
```suggestion
        optimizer = torch.optim.Adam(self.model.parameters(), lr=self.learning_rate)
```

File: tutorials/train_classifier.ipynb [hunk 27]
```diff
@@ -0,0 +1,248 @@
+{
```
Comment (Pringled) [inline tutorials/train_classifier.ipynb:1]: This prints in the notebook are not saved/pushed I think. This is a bit weird, for example when you have:
"Pretty good! We outperform the tf-idf pipeline by a wide margin.", it doesn't actually show it in the notebook (see https://github.com/MinishLab/model2vec/blob/add-fittable/tutorials/train_classifier.ipynb).

**Review by Pringled — CHANGES_REQUESTED** [review]
Couple of comments and suggestions

**Review by Pringled — APPROVED** [review]
:heart_eyes:
